src/distances/qgram.jl

```diff
@@ -80,9 +78,141 @@ function _count(s1, s2)
 	return values(d)
 end
 
+# Turn a sequence of qgrams to a count dict for them, i.e. map each
+# qgram to the number of times it has been seen.
+function countdict(qgrams)
+    d = Dict{eltype(qgrams), Int32}()
+    for qg in qgrams
+        index = Base.ht_keyindex2!(d, qg)
+		if index > 0
+			d.age += 1
+			@inbounds d.keys[index] = qg
+			@inbounds d.vals[index] = d.vals[index][1] + 1
+		else
+			@inbounds Base._setindex!(d, 1, qg, -index)
+		end
+    end
+    d
+end
+
+abstract type AbstractQGramCounts{Q,K} end
+q(qc::AbstractQGramCounts{Q,K}) where {Q,K} = Q
+counts(qc::AbstractQGramCounts) = qc.counts
+
+struct QGramDict{Q,K} <: AbstractQGramCounts{Q,K}
```

Comment:
Do you really want to have both QGramDict and QGramSortedArray? Would not it be simpler to only have QGramSortedArray?

Reply:
Yes, I would very strongly prefer to keep both in there. At least two benefits to the QGramDict one:
1. We can support streaming QGram counting (think very large strings/streams),
2. We can do set "arithmetics" on QGramDicts for specific distances or when adding Diversity/Diameter distances that work on sets/vectors of strings.

I would argue that the benefits outweigh the downside of having to support also QGramDict. The code for it is limited and simple so shouldn't be a problem.

If anything, I would propose extracting the iterators and QGram dict/sortedvector code to a qgram_helpers.jl or qgram_counting.jl file to keep things more modular and separate. This latter file would be more nitty-gritty and not needed for people that just want to look at the QGramDistances and the code that implements them.